Add Utah teapot as a geometry object; doc cleanup

erich666 opened the pull request:
Add the Utah teapot geometry object, along with a demo showing its
options. Also fix Matrix\* documentation: match current constructor code,
and clean up style.

This geometry object could be integrated into src/extras/geometries, and I'm happy to make the corresponding HTML documentation page. My hesitation with asking for this object to be added as a "first class" geometry object that gets included into three.min.js is just that it has some large arrays in it. Most geometry objects are fairly small (ExtrudeGeometry.js being the exception), so I am a little shy to recommend TeapotGeometry.js as a true part of three.js. Do note that if TeapotGeometry.js is minified it is about 10k, down from 17k.

erich666 committed:
Add Utah teapot as a geometry object; doc cleanup

Add the Utah teapot geometry object, along with a demo showing its
options. Also fix Matrix* documentation: match current constructor code,
and clean up style.

WestLangley commented on examples/js/geometries/TeapotGeometry.js:
```diff
@@ -0,0 +1,634 @@
+					if ( notDegenerate ( this.vertices[v1], this.vertices[v2], this.vertices[v3] ) ) {
+						this.faces.push( new THREE.Face3( v1, v2, v3, [ normals[v1], normals[v2], normals[v3] ] ) );
+						this.faceVertexUvs[ 0 ].push( [ uvs[v1], uvs[v2], uvs[v3] ] );
+					}
```
```javascript
this.faces.push( new THREE.Face3( v1, v2, v3, [ normals[ v1 ].clone(), normals[ v2 ].clone(), normals[ v3 ].clone() ] ) );
this.faceVertexUvs[ 0 ].push( [ uvs[ v1 ].clone(), uvs[ v2 ].clone(), uvs[ v3 ].clone() ] );
```
Normals and UVs cannot be shared. You can see the consequences of sharing if you call `geometry.applyMatrix( matrix )`.

You can use [mrdoobapproves](http://zz85.github.io/mrdoobapproves/) if you want help with your formatting.

WestLangley commented on examples/webgl_geometry_teapot.html:
```diff
@@ -0,0 +1,383 @@
+				phongMaterial.shininess = effectController.shininess;
+				texturedMaterial.shininess = effectController.shininess;
+
+				var diffuseColor = new THREE.Color();
```
I'd avoid instantiating a new `THREE.Color` inside the render loop. Create one and reuse it.

WestLangley commented:
Thanks! The demo is great, as are the doc improvements.

I am a bit confused, however, why your gui API uses the `ka / kd / ks` parameterization for the 'material control'. three.js does not use that nomenclature.

erich666 commented:
Thanks for the great (and quick) comments. I'll fix the Three.color for sure. I should change Ka, Kd, and Ks - those are "traditional" and used in the Udacity course, but I should move to Three terminology.

"Normals and UVs cannot be shared." - oh, interesting! I'll clone() my little heart out. I'll also run the whole thing through mrdoobapproves.

erich666 committed:
Fixes from West, and  mrdoobapproves cleanups

I won't reformat "-5.3" to be "- 5.3", etc. I didn't bother adding
spaces in the giant arrays of spline surface data.

erich666 commented:
OK, I've updated the code with a new commit - let me know if you have other comments.

mrdoob commented:
Yay Teapot! 🎉

mrdoob commented:
> This geometry object could be integrated into src/extras/geometries, and I'm happy to make the corresponding HTML documentation page. My hesitation with asking for this object to be added as a "first class" geometry object that gets included into three.min.js is just that it has some large arrays in it. 

Yeah, I think I would put it on `examples/js/geometries/` instead.

mrdoob commented on examples/js/geometries/TeapotGeometry.js:
```diff
@@ -0,0 +1,675 @@
+
+	this.type = 'TeapotGeometry';
+
+	this.size = size || 50;
```
Instead of creating public properties you should save them in a `parameters` object. You can use [CircleGeometry](https://github.com/mrdoob/three.js/blob/dev/src/extras/geometries/CircleGeometry.js) as reference.

erich666 committed:
Use parameter setup

Done similar to what CircleGeometry does.

erich666 commented:
OK, I've rewritten the parameter code to follow the form of CircleGeometry.

> Yeah, I think I would put it on examples/js/geometries/ instead.

Sounds good. It's well-documented at the top of the file anyway, so no HTML page is no great loss.

I considered making a more artsy teapot demo, and maybe I will at some point ([Mad Hatter's tea party](https://en.wikipedia.org/wiki/The_Hatter) as an inspiration), but thought I'd start small and simple, with clean demo code for how to use the teapot.

WestLangley commented on examples/webgl_geometry_teapot.html:
```diff
@@ -0,0 +1,401 @@
+					fragmentShader: shader.fragmentShader,
+					vertexShader: shader.vertexShader,
+					uniforms: shader.uniforms,
+					depthWrite: false,
```
`depthWrite: false` should not be needed

WestLangley commented on examples/webgl_geometry_teapot.html:
```diff
@@ -0,0 +1,401 @@
+
+			function render() {
+
+				var delta = clock.getDelta();
```
`delta` is not needed.

WestLangley commented on examples/webgl_geometry_teapot.html:
```diff
@@ -0,0 +1,401 @@
+
+				var delta = clock.getDelta();
+
+				cameraControls.update( delta );
```
`controls.update()` is not needed.

WestLangley commented:
Nice demo!

Your flat shading is not rendering `flat`. It would if you switched to `MeshPhongMaterial`. I am not sure what is going on with `Lambert` in this case -- there have been so many changes...

Also, you are not disposing of memory properly. Type `renderer.info` into the console and you will see the number of geometries growing without bound.

tschw commented on examples/js/geometries/TeapotGeometry.js:
```diff
@@ -0,0 +1,701 @@
+					vtp = new THREE.Vector4( tp[ 0 ], tp[ 1 ], tp[ 2 ], tp[ 3 ] );
+					vdsp = new THREE.Vector4( dsp[ 0 ], dsp[ 1 ], dsp[ 2 ], dsp[ 3 ] );
+					vdtp = new THREE.Vector4( dtp[ 0 ], dtp[ 1 ], dtp[ 2 ], dtp[ 3 ] );
+
```
Those could be allocated outside of the loop and `fromArray`ed...

tschw commented on examples/js/geometries/TeapotGeometry.js:
```diff
@@ -0,0 +1,701 @@
+
+						// multiply power vectors times matrix to get value
+						tcoord = vsp.clone();
+						tcoord.applyMatrix4( mgm[ i ] );
```
Instead of creating clones could `copy` into preallocated vectors.

tschw commented:
I think we absolutely need "Add -> Teapot" in the pull down menu of the editor! It's great to always have a nontrivial object aboard for testing / debugging.

![solids_rendering](https://cloud.githubusercontent.com/assets/3916357/9566708/2db0c55c-4f10-11e5-92c0-f8d1525897d4.jpg)

Since it computes all the data itself, seems it could easily become a `TeapotBufferGeometry` - directly writing the data to flat arrays...

mrdoob commented:
> I am not sure what is going on with `Lambert` in this case -- there have been so many changes...

It's because at the moment `THREE.FlatShading` is automatically computed in the fragment shader (using derivatives).

It's a problem with attributes. At the moment `Geometry` -> `BufferGeometry` generates this:

```
{ "position", "normal", "color", "uv" }
```

Because the same geometry can be rendered with different materials there are things (ie `VertexColor`/`FaceColor`, `SmoothShading`/`FlatShading`, ...) that require different attributes.

So, need to find a way to generate all the possible attribute buffers. One option would be to generate this;

```
{ "position", "vertexNormal", "faceNormal", "vertexColor", "faceColor", "uv" }
```

Then, at render time, upload one or the other depending of the material parameters.

But, at the same time, those parameters don't work with `BufferGeometry`.

mrdoob commented:
> I think we absolutely need "Add -> Teapot" in the pull down menu of the editor! 

Agreed!

> Since it computes all the data itself, seems it could easily become a `TeapotBufferGeometry` - directly writing to flat arrays...

Also agreed!

WestLangley commented:
> It's because at the moment THREE.FlatShading is automatically computed in the fragment shader (using derivatives).

Yes. That is how `MeshPhongMaterial` does it.

Let's decide. Do you want to move `MeshLambertMaterial` out of the library? I you suggested that recently in #6586.

mrdoob commented:
> Let's decide. Do you want to move `MeshLambertMaterial` out of the library? I you suggested that recently in #6586.

Well, that's one of the options. It'll probably happen but I wanted to wait.

Another way of looking at it is to question whether `THREE.SmoothShading`/`THREE.FlatShading` and `THREE.FaceColors`/`THREE.VertexColors` should be in the material.

WestLangley commented:
@mrdoob Hmmm... Your questions are very good ones. In fact, they deserve their own thread IMO, so we don't hijack this one.

erich666 committed:
Teapot fixes, change to BufferGeometry

Changed to BufferGeometry object, avoid a number of unnecessary allocs,
flat face cleanup.

erich666 commented:
Wonderful review comments, all!

> depthWrite: false should not be needed

It's an optimization: the skybox is drawn first. It does not need z-buffering. Turning the z-buffer write I believe leads to less memory access (doesn't read or write the z-buffer), so should be a tiny bit faster. This is used in a number of examples already (which is where I got it), e.g., webgl_materials_envmaps.html. If not good here, those should change, too.

> delta is not needed.
> controls.update() is not needed.
> Actually, this is a static scene, so an animation loop is not needed either.

West, could you suggest alternate code? I'm so used to "function animate()" that I don't know what to use instead.

> Your flat shading is not rendering flat. It would if you switched to MeshPhongMaterial. I am not sure what is going on with Lambert in this case -- there have been so many changes...

Great catch! Funny, it used to work. Anyway, switched to PhongMaterial, which is certainly overkill for flat shading. Anyway, I see your whole discussion; best of luck. The code works now, so I'll leave it as-is and you guys can change it later if you want (the Lambert shader has to be a bit faster than the Phong shader).

> Also, you are not disposing of memory properly. Type renderer.info into the console and you will see the number of geometries growing without bound.

I think I cleaned this up. I do see Array[1] grow to as much as Array[7] now, but no further - those are the shaders created along the way, and there's a maximum of 7.

> Those could be allocated outside of the loop and fromArrayed...

Nice! I didn't know about fromArray - done. Also moved two matrices and other stuff to be allocated just once.

> Instead of creating clones could copy into preallocated vectors.

I don't really understand the whole "clone()" thing for the normals and UVs as it is. See next comment...

> Since it computes all the data itself, seems it could easily become a TeapotBufferGeometry - directly writing to flat arrays...

That's a good idea, and I should have gone that route from the start! Converted over, and amazingly, it ran the first time. It's also super-fast to generate. Give it a look, and check the renames.

And, yes, the teapotahedron is the sixth Platonic solid, as you show in Jim Arvo's famous image. I used this image in my video about the teapot, too: https://www.youtube.com/watch?v=DxMfblPzFNc - 3DS Max has the teapot as a basic primitive (for reasons of having something interesting to render), and I use it throughout the Udacity MOOC for this reason. Adding it to the editor is a good plan.

WestLangley commented:
You have a static scene, and an animation loop is not needed.

1. remove `animate()` and call `render()` instead
2. remove `stats`
3. add `cameraControls.addEventListener( 'change', render );`
4. add `.onChange( render ); ` to each gui element

WestLangley commented:
`renderer.info.memory.geometries` is still growing without bound as you re-tessellate.

See http://threejs.org/examples/webgl_test_memory.html.

erich666 committed:
Cleanups: fix memory leak, update only on changes

Removed stats and animation loop, since scene needs update only on
input. Fixed mesh memory leak.

erich666 commented on examples/webgl_buffergeometry_teapot.html:
```diff
@@ -0,0 +1,391 @@
+				
+				if ( teapotMesh !== undefined ) {
+
+					teapotMesh.dispose();
```
West: OK, I got the switchover from animate to render to work and tested it thoroughly.

I added the "dispose()" for the teapot's mesh geometry object. I honestly don't understand why this dispose() is needed. Is this a flaw with three.js? See fillScene(). Whenever I make a new teapot, I make a new scene rather than remove the teapot and skybox from the old scene and add back into it. Shouldn't the old scene simply "go away" at this point, along with all its objects and their meshes? I would have thought that garbage collection would note that the mesh (now called teapotMesh, so I can dispose() it manually) was no longer used by anything. Does anyone know what holds onto teapotMesh such that I have to call dispose() on it?

mrdoob commented:
> Does anyone know what holds onto teapotMesh such that I have to call dispose() on it?

The issue is that buffer were created and uploaded to the GPU. In JavaScript there is no way to know when an object has been GCed so we can't delete those buffers automatically.

mrdoob commented:
Maybe in the future we can use [WeakMap](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap) for this though

erich666 commented:
Thanks, this whole conversational thread has been educational for me, I appreciate all the work that you guys put into it. Code's much better (I think) and teapot generation is now super-fast and efficient, display doesn't waste cycles, and all the rest.

mrdoob commented on examples/webgl_buffergeometry_teapot.html:
```diff
@@ -0,0 +1,391 @@
+			// Whenever the teapot changes, the scene is rebuilt from scratch (not much to it).
+			function fillScene() {
+
+				scene = new THREE.Scene();
```
One final change... I don't think it's good practice to create a scene for every change.

I think it'd be better if the lights were added just once at init time and then do this every change:

```javascript
if ( teapot !== undefined ) {
	teapot.geometry.dispose();
	scene.remove( teapot );
}

...
```

WestLangley commented on examples/webgl_buffergeometry_teapot.html:
```diff
@@ -0,0 +1,391 @@
+					
+				}
+
+				teapotMesh = new THREE.TeapotBufferGeometry( teapotSize,
```
`teapotGeometry`, not `teapotMesh`

erich666 committed:
Don't destroy scene, just repopulate it

Show suggested best practice for managing a scene.

erich666 committed:
teapotMesh to teapotGeometry

name change

erich666 commented:
OK, I changed the scene to be generated once and repopulated as needed. Also, teapotMesh to teapotGeometry rename, which I noticed the comment on right after syncing.

mrdoob commented:
Many thanks! 😊
